File: arrays/keysAndRooms.cpp
// Leetcode - Keys and Rooms

void utility(int roomNo, vector<int> &visited, vector<vector<int>> &rooms)
{
    visited[roomNo] = 1;

    for (auto key : rooms[roomNo])
    {
        if (!visited[key])
        {
            utility(key, visited, rooms);
        }
    }
}

bool canVisitAllRooms(vector<vector<int>> &rooms)
{
    vector<int> visited(rooms.size(), {0});
    utility(0, visited, rooms);

    for (int i = 0; i < visited.size(); i++)
    {
        if (visited[i] == 0)
        {
            return false;
        }
    }
    return true;
}

File: dfs/keysAndRooms.java
// Leetcode - 841 - Keys and Rooms

public boolean canVisitAllRooms(List<List<Integer>> rooms) {
    HashSet<Integer> visited = new HashSet<>();
    visited.add(0);
    
    Stack<Integer> stack = new Stack<>();
    stack.add(0);
    
    while(!stack.isEmpty()) {
        List<Integer> keys = rooms.get(stack.pop());
        for(int key: keys) {
            if(!visited.contains(key)) {
                visited.add(key);
                stack.add(key);
            }
        }
    }
    return visited.size() == rooms.size();
}